packages/mock/src/fulcio/ca.ts

```diff
@@ -0,0 +1,181 @@
+/*
+Copyright 2023 The Sigstore Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { Crypto } from '@peculiar/webcrypto';
+import x509 from '@peculiar/x509';
+import * as asn1js from 'asn1js';
+import {
+  DIGEST_SHA256,
+  KEY_ALGORITHM_ECDSA_P256,
+  KEY_ALGORITHM_ECDSA_P384,
+  SIGNING_ALGORITHM_ECDSA_SHA384,
+} from '../constants';
+import { createRootCertificate } from '../util/root-cert';
+import type { CTLog } from './ctlog';
+
+const ISSUER = 'CN=sigstore,O=sigstore.mock';
+
+const OID_SIGNED_CERTIFICATE_TIMESTAMP_LIST = '1.3.6.1.4.1.11129.2.4.2';
+
+const MS_PER_MINUTE = 1000 * 60;
+
+export interface CA {
+  rootCertificate: Buffer;
+  issueCertificate: (options: CertificateRequestOptions) => Promise<Buffer>;
+}
+
+export interface CertificateRequestOptions {
+  publicKey: Buffer;
+  subjectAltName: string;
+  extensions?: ExtensionValue[];
+}
+
+interface ExtensionValue {
+  oid: string;
+  value: string;
+}
+
+// Initialize the mock CA. Starts by generating a root key pair and self-signed
+// certificate. The root certificate is used to sign any certificates issued by
+// the #issueCertificate method. Doing this work in a function allows us to
+// work around the fact that some of these operations are async and can't be
+// done in the constructor.
+export async function initializeCA(ctLog?: CTLog): Promise<CA> {
+  const root = await createRootCertificate(
+    ISSUER,
+    KEY_ALGORITHM_ECDSA_P384,
+    SIGNING_ALGORITHM_ECDSA_SHA384
+  );
+
+  return new CAImpl({
+    keyPair: root.keyPair,
+    rootCertificate: root.cert,
+    ctLog,
+  });
+}
+
+class CAImpl implements CA {
+  private root: x509.X509Certificate;
+  private keyPair: CryptoKeyPair;
+  private ctLog?: CTLog;
+  private crypto: Crypto;
+
+  constructor(options: {
+    rootCertificate: x509.X509Certificate;
+    keyPair: CryptoKeyPair;
+    ctLog?: CTLog;
+  }) {
+    this.root = options.rootCertificate;
+    this.ctLog = options.ctLog;
+    this.keyPair = options.keyPair;
+    this.crypto = new Crypto();
+  }
+
+  public get rootCertificate(): Buffer {
+    return Buffer.from(this.root.rawData);
+  }
+
+  public async issueCertificate(
+    opts: CertificateRequestOptions
+  ): Promise<Buffer> {
+    // convert key to CryptoKey
+    const key = await this.crypto.subtle.importKey(
+      'spki',
+      opts.publicKey,
+      KEY_ALGORITHM_ECDSA_P256,
+      true,
+      ['sign', 'verify']
+    );
+
+    const tbs = {
```

Comment:
Anything useful we could link to in fulcio for the actual implementation for ref?